### PR TITLE
Ensure Messaging instance is usable after FIRApp's `delete`

### LINKED ISSUE #3411: Using FirebaseApp.delete() results in unusable Messaging singleton
### Describe your environment

  * Xcode version: 10.2.1
  * Firebase SDK version: 6.5.0
  * Firebase Component: Messaging
  * Component version: 4.1.1

### Describe the problem
We have several projects set up, representing different environments for remote config. We use Messaging to enable real time remote config updates. When a user switches the firebase environment within the app, we delete the default app, and attempt to recreate it.

```
FirebaseApp.delete({ _ in })
FirebaseApp.configure(options: options)
```

Once this has been done any interaction with Messaging will not work. We have encountered this because we were running a very old version of Firebase (4.13.0) & Messaging (2.2.0) and we're trying to update to the latest. This approach worked in those versions.

This problem looks to be in the implementation of `messaging()`, there is a `dispatch_once` block to actually set everything up, this doesn't make sense if the singleton object can actually be deallocated and recreated.

I have attempted to solve this by calling `start` manually, and this seems to fix some things, but the new Messaging instance never gets notified of the `fcmToken` being available so cannot subscribe to anything.

### PR DESCRIPTION
**Pull request by ryanwilson**
This fixes #3411. Test fails before the code change, and succeeds
afterwards.

### CODE REVIEW

File: Example/Messaging/Tests/FIRMessagingInstanceTest.swift [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import FirebaseCore
+import FirebaseMessaging
+import XCTest
+
+class FIRMessagingInstanceTest: XCTestCase {
```
Comment (maksymmalyhin) [inline Example/Messaging/Tests/FIRMessagingInstanceTest.swift]: Nice, tests in Swift!

File: FirebaseMessaging.podspec [hunk 2]
```diff
@@ -52,7 +52,7 @@ device, and it is completely free.
   s.dependency 'Protobuf', '~> 3.1'
 
   s.test_spec 'unit' do |unit_tests|
-    unit_tests.source_files = 'Example/Messaging/Tests/*.[mh]'
+    unit_tests.source_files = 'Example/Messaging/Tests/*.{m,h,swift}'
```
Comment (paulb777) [inline FirebaseMessaging.podspec]: Awesome! swift unit tests!

**Comment by paulb777** [comment]
Add to release notes and bug to milestone if this fix is going in M54 (6.7.0) cc: @chliangGoogle

**Comment by ryanwilson** [comment]
@chliangGoogle does this make sense that `setup` then `teardown` and `setup` could be called again, or will this case issues?

**Comment by ryanwilson** [comment]
> Can you also update the CHANGELOG if this goes into M54 release?

Done!

File: Firebase/Messaging/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # 2019-08-20 -- v4.1.3
 - [changed] Cleaned up the documents, unused macros, and folders. (#3490, #3537, #3556, #3498)
 - [changed] Updated the header path to pod repo relative. (#3527)
+- [fixed] Fix singleton functionality after a FirebaseApp is deleted and recreated. (#3579)
```
Comment (paulb777) [inline Firebase/Messaging/CHANGELOG.md]: Use bug number - #3411

Comment (ryanwilson) [reply Firebase/Messaging/CHANGELOG.md]: Good catch!

**Comment by paulb777** [comment]
Building the swift file fails with the --use-libraries test in the cron job. https://travis-ci.org/firebase/firebase-ios-sdk/jobs/572152503

**Pull request merged**